GEN-1632 factories deployments

fedealconada opened the pull request:
## Motivation

Deploy fixed 4626 factories
<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution

- Deploys fixed 4626 factories
- I've also added the task for the re-deployment of the Periphery (thought it my be scheduled for later)  
<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

## Implementer Checklist

<!--
Fill out the checklist below as applicable to your change. Bug fixes and new features must 
check off all of the required items.
-->

- [ ]  [OPTIONAL] Create a reference implementation (in python or JS)
- [x]  Create a short "Motivation" section for the PR
- [x]  Write a spec for the feature and put it in the PR description – basic function names and expected state transitions is OK
- [ ]  Get the PR reviewed by at least two people

If this is your first time reviewing smart contract changes
- [ ]  Review the [Solcurity](https://github.com/Rari-Capital/solcurity) standard

If smart contract changes were made
- [ ]  Check all of the new revert paths with concrete tests
- [ ]  Check all of the new storage slot writes with concrete tests – do not make the passing test case the trivial case!
- [ ]  Go line-by-line through the new code and ensure it has all been covered by concrete tests – since we don’t have a coverage engine for foundry yet, you can imagine how one might check that each different code path is covered
- [ ]  Simplify the implementation and spend some time trying to minimize gas costs
- [ ]  Write fuzz tests for the feature – try everything to break the implementation (is this function monotonically in/decreasing, should it always be less than something else, etc)
- [ ]  Capture bugs discovered during the above steps in concrete tests (regression tests)
- [ ]  Add integration tests – fuzz or concrete tests that test how the new code affects the entire system, either locally or with a mainnet fork
- [ ]  [OPTIONAL] Integrate the feature into the deployment scripts
- [ ]  [OPTIONAL] Add new sanity checks to the deployment scripts

If your PR uses or interacts with prices of any kind
- [ ] Make sure forge is running the test on the latest block
- [ ] Compare the value returned with some other source (like Curve)
- [ ] Check the `updatedAt` (if available) value of the contract and confirm that the price we are sourcing is in fact not older than X
- [ ] Ensure oracle liveness from the oracle maintainer, either through a public status interface or through written confirmation from a developer.

fedealconada committed:
chore: add deployment task for factories

fedealconada committed:
chore: add deployment task for new periphery

fedealconada committed:
chore: appease linter

jparklev commented on pkg/deployments/tasks/20221028-factories-and-morpho/input.js:
```diff
@@ -0,0 +1,84 @@
+    targets: SAMPLE_TARGETS,
+  },
+  {
+    contractName: "@sense-finance/v1-core/src/adapters/abstract/factories/ERC4626Factory.sol:ERC4626Factory",
```
why is this a path?

fedealconada replied:
should have clarified! since there's a contract from yield daddy called `ERC4626Factory`, it conflicts with our own `ERC4626Factory` so hardhat makes you specify which exactly you want to usse

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,170 @@
+      await newPeriphery.onboardAdapter(adapter, false).then(t => t.wait());
+    }
+
+    for (let adapter of adaptersOnboarded) {
```
I believe we'll have to do this with factories, too (add factories that are on the existing periphery). I don't know if we had during the previous periphery upgrade?

jparklev replied:
Also don't remember if we've had to remove adapters and if we should filter those out of this list somehow?

fedealconada replied:
ohh, you're right, good call, will add it

jparklev commented on pkg/deployments/tasks/20221028-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,188 @@
+          console.log(`\n- Onboard target ${t.name} @ ${t.address} via ${factoryContractName}`);
+          const data = ethers.utils.defaultAbiCoder.encode(["address[]"], [[]]);
+          const adapterAddress = await periphery.callStatic.deployAdapter(factoryAddress, t.address, data);
+          await (await periphery.deployAdapter(factoryAddress, t.address, data)).wait();
```
So this fails with USDT rn, right?

fedealconada replied:
yep

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,170 @@
+      console.log("\nGive the new Periphery auth over the pool manager");
+      await (await poolManager.setIsTrusted(peripheryAddress, true)).wait();
+
+      // TOOD: we don't need this right? It was on the '20220313-periphery' but don't remember why?
```
I believe that's right. I think we tried to batch this in with the last update, but it was extra

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/input.js:
```diff
@@ -0,0 +1,10 @@
+    divider: "0x86bA3E96Be68563E41c2f5769F1AF9fAf758e6E0",
+    periphery: "0xFff11417a58781D3C72083CB45EF54d79Cd02437",
+    spaceFactory: "0x5f6e8e9C888760856e22057CBc81dD9e0494aA34",
+    poolManager: "0xf01eb98de53ed964AC3F786b80ED8ce33f05F417",
```
This should be the noop pool manager from task "noop-pool-manager", right? (`0x9BFcA4882c1D8CDE3b4f787d3d74eb0E1bb0ec1B`)

fedealconada replied:
Ohh, yes! I actually took this address from the Observable thinking it was the last version, we should update it there too!

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,170 @@
+      const multisigSigner = await hre.ethers.getSigner(senseAdminMultisigAddress);
+      oldPeriphery = oldPeriphery.connect(multisigSigner);
+      divider = divider.connect(multisigSigner);
+      poolManager = poolManager.connect(multisigSigner);
```
since we're using the noop pool manager, we can probably get rid of all of the pool manager calls here

jparklev reviewed (commented):
A few quick comments before executing ~

fedealconada committed:
chore: re-add factories on Periphery re-deployment, update pool manager address, addrerss issues from review

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -71,6 +69,28 @@ task("20221029-LATER-periphery", "Deploys and authenticates a new Periphery").se
+        (await oldPeriphery.queryFilter(oldPeriphery.filters.FactoryChanged(null))).map(e => e.args.factory),
+      ),
+    ];
+    console.log("Factories to verify:", factoriesVerified);
```
Should we be doing something with these?

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -50,7 +48,7 @@ task("20221029-LATER-periphery", "Deploys and authenticates a new Periphery").se
       ),
     ];
-    console.log("Adapters to onborad:", adaptersOnboarded);
+    console.log("Adapters to onboard:", adaptersOnboarded);
```
Have we removed any adapters from the divider? do we need to filter those out somehow?

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -71,6 +69,28 @@ task("20221029-LATER-periphery", "Deploys and authenticates a new Periphery").se
+
+    const factoriesVerified = [
+      ...new Set(
+        (await oldPeriphery.queryFilter(oldPeriphery.filters.FactoryChanged(null))).map(e => e.args.factory),
```
should we filter out factories that we've unverified?

jparklev reviewed (commented):
A few last qs for clarity on verification and unsetting adapters ~

fedealconada committed:
chore: comment crop(s) factories from deployment script

fedealconada committed:
chore: appease linter

fedealconada committed:
chore: fix tests

fedealconada committed:
chore: fix file name

fedealconada committed:
chore: add ABIs

abylaw commented on pkg/deployments/tasks/index.js:
```diff
@@ -15,3 +15,5 @@ require("./20220922-4626-crop-factory");
 require("./20220926-euler");
 require("./20221007-morpho-adapters");
+require("./20221031-factories-and-morpho");
+require("./20221029-LATER-periphery");
```
should we call this task `periphery-v1.4 (i think???)` instead? more context :)

abylaw commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,144 @@
+      await newPeriphery.onboardAdapter(adapter, false).then(t => t.wait());
+    }
+
+    for (let adapter of adaptersOnboarded) {
```
```suggestion
    for (let adapter of adaptersVerified) {
```
?

abylaw commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,144 @@
+    console.log("\nRemoving deployer as admin on Periphery");
+    await newPeriphery.setIsTrusted(deployer, false).then(t => t.wait());
+
+    // If we're in a forked environment, check the txs we need to send from the multisig as well
```
meganit: move comment down into `else`?

abylaw commented on pkg/deployments/tasks/20221029-LATER-periphery/index.js:
```diff
@@ -0,0 +1,144 @@
+          to: senseAdminMultisigAddress,
+          value: ethers.utils.parseEther("1"),
+        })
+      ).wait();
```
`.then(t => t.wait())` to be consistent?

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,189 @@
+      );
+
+      const factoryParams = [masterOracleAddress, stake, stakeSize, minm, maxm, ifee, mode, tilt, guard];
+      const { address: factoryAddress, abi } = await deploy("ERC4626Factory_Sense", {
```
does this name get surfaced anywhere? should it be `factoryContractName`?

fedealconada replied:
i've removed this! it was temporary while I was trying some things, sorry!

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,189 @@
+
+      const factoryParams = [masterOracleAddress, stake, stakeSize, minm, maxm, ifee, mode, tilt, guard];
+      const { address: factoryAddress, abi } = await deploy("ERC4626Factory_Sense", {
+        contract: "@sense-finance/v1-core/src/adapters/abstract/factories/ERC4626Factory.sol:ERC4626Factory",
```
oh! interesting syntax. does it refer to a specific contract within that file?

fedealconada replied:
will be removed from here, but we need this notation to specify exactly which contract we are referring to since there's another contract with that exact name from a lib. we are using (yield-daddy)
I've added a comment about this :)

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,189 @@
+        console.log(`\nDeploy adapters for: ${factoryContractName}`);
+        console.log("\n-------------------------------------------------------");
+
+        for (const t of targets) {
```
wonder if we should pull out deploying adapters for factories into a separate task.

we'd reduce code duplication by quite a bit, iic.
so we'd run:
1) `20221031-factories-and-morpho`
2) `deploy-adapters`

or maybe pull adapters into a function?

fedealconada replied:
to be honest, this happens on hardhat fork only, as is more like a sanity check when deploying a factory, just to make sure we can deploy an adapter for t and sponsor a series. In mainnet or goerli, we would deploy the adapter using OZ Defender.

fedealconada replied:
but, we could definitely extract this out (maybe on a utils file) because it seems like something we are constantly using and it's copied over and over from one task to another one (i'll mark is as a quality work)

abylaw replied:
great ok! makes sense~

fedealconada committed:
chore: fix etherscan verification and address comments from PR review

fedealconada commented:
Oka, re-bumping this since the etherscan verification issue blocked my for a while.
Would need a re-review from you both @abylaw and @jparklev 🙏🏻
- I've addressed your last comments on the PR (@jparklev I'm filtering non-used adapters and factories for the Periphery re-deployment)
- I've fixed the issue with the verification. Context: seems like `hardhat-preprocessor` plugin conflicts we the verification task we were using (which, somehow, was not the one built-in with `hardhat-deploy`. Switching to that one worked well but with some caveats:
  - It makes me use the `solcInput` which, according to the docs, some time was needed as a workaround for a bug that was affecting verification for contracts using Solidity version < 0.8. Since we are using >0.8 we should not need this but somehow it only works with this param.
  - Also, after successfully verifying, when going to Etherscan I was seeing a warning saying that a library called `__CACHE_BREAKER` was not verified. According to `https://github.com/wighawag/hardhat-deploy/issues/78#issuecomment-786914537` this is something added when using `hardhat-preprocessor` but it's safe to remove. So, I've added a little workaround that, before submitting for verifying, I'm removing this library from the `solcInput` file. 

I've already asked about these 2 things [here](https://github.com/wighawag/hardhat-deploy/issues/78) and [here](https://github.com/wighawag/hardhat-preprocessor/issues/12#issuecomment-1310159980)

fedealconada committed:
chore: modify fix to work with anyn network

jparklev commented on pkg/deployments/hardhat.utils.js:
```diff
@@ -128,22 +128,35 @@ exports.setStorageAt = setStorageAt;
+
+  // Remove __CACHE_BREAKER library from solcInput
+  const path = `${__dirname}/deployments/${hre.network.name}/solcInputs`;
+  fs.readdirSync(path).forEach(file => {
```
Do you know why this wasn't appearing for us before? I suppose we'll wait for the responses on https://github.com/wighawag/hardhat-deploy/issues/78?

fedealconada replied:
It was not because we are using a different verification task and the one built-in on the `hardhat-deploy` plugin. And seems like when using the preprocessor we should be using the `hardhat-deploy` built-in verification task.
Before, it was working because we were not making use of the preprocessor

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery-v1.4/index.js:
```diff
@@ -0,0 +1,131 @@
+
+const { verifyOnEtherscan } = require("../../hardhat.utils");
+
+task("20221029-LATER-periphery-v1.4", "Deploys and authenticates a new Periphery").setAction(
```
Echo @abylaw's suggestion here! <img width="810" alt="Screen Shot 2022-11-10 at 4 02 11 PM" src="https://user-images.githubusercontent.com/24902242/201205138-8c3764c8-831d-42c3-8b04-3956a630fbeb.png">

abylaw replied:
also +1 on removing `later` :) we know it's for later!

jparklev commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,185 @@
+          restrictedAdmin,
+          rewardsRecipient,
+          factoryParams,
+          ...(factoryContractName === "ERC4626CropFactory" ? [ethers.constants.AddressZero] : []),
```
why is the reward token the zero address?

abylaw replied:
i also wish to know...

fedealconada replied:
To be honest, it's good that you are asking. The `ERC4626CropFactory.sol` inherits `CropFactory` which requires a `reward` on the constructor. But, actually, on `ERC4626CropFactory` we should not require `reward` on the constructor because we will never use that (we will never have a common reward token across ERC4626 tokens haha).
So, I've just removed that param from the constructor since it was useless!

abylaw replied:
Could we leave a comment so next person reading is not surprised? :)

fedealconada replied:
Hm.. but this is gone now :)

jparklev commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,185 @@
+      const factoryParams = [masterOracleAddress, stake, stakeSize, minm, maxm, ifee, mode, tilt, guard];
+
+      const { address: factoryAddress, abi } = await deploy(factoryContractName, {
+        contract: !contract ? factoryContractName : contract,
```
```suggestion
        contract: contract ? contract : factoryContractName,
```
#nit

fedealconada replied:
yep, we can actually do `contract || factoryContractName`

jparklev commented on pkg/deployments/tasks/20221029-LATER-periphery-v1.4/index.js:
```diff
@@ -0,0 +1,131 @@
+    for (let adapter of adaptersOnboarded) {
+      console.log("\nOnboarding adapter", adapter);
+      const params = await divider.adapterMeta(adapter);
+      if (parseInt(params[2]) > 0) {
```
could potentially do something like `params[2].gt(0)` or smtg here as well

fedealconada replied:
done! i thought it was a string

jparklev reviewed (approved):
Didn't go as deep on the hardhat testing stuff, but deployment code LGTM 🦾

abylaw commented on pkg/deployments/hardhat.utils.js:
```diff
@@ -128,22 +128,35 @@ exports.setStorageAt = setStorageAt;
+    fs.writeFileSync(`${path}/${file}`, JSON.stringify(m));
   });
+
+  // Try verification
```
nit: remove comment

abylaw commented on pkg/deployments/hardhat.utils.js:
```diff
@@ -128,22 +128,35 @@ exports.setStorageAt = setStorageAt;
+  // Remove __CACHE_BREAKER library from solcInput
+  const path = `${__dirname}/deployments/${hre.network.name}/solcInputs`;
+  fs.readdirSync(path).forEach(file => {
+    console.log(file);
```
is this for debugging or is there something going on here i don't understand? :) 
if for debugging, we should remove it.

fedealconada replied:
nice catch hehe, it was just me debugging

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,185 @@
+          })
+        ).wait();
+
+        // impersonate multisig account
```
nit: remove comment

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,185 @@
+        console.log("\n-------------------------------------------------------");
+        await verifyOnEtherscan(factoryContractName);
+      } else {
+        // fund multisig
```
nit: remove comment. `console.log` is enough

abylaw commented on pkg/deployments/tasks/20221031-factories-and-morpho/index.js:
```diff
@@ -0,0 +1,185 @@
+
+      console.log(`${factoryContractName} deployed to ${factoryAddress}`);
+
+      // if not hardhat fork, we try verifying on etherscan
```
nit: remove comment

abylaw reviewed (commented):
Yay!!

fedealconada committed:
chore: address comments from PR

fedealconada committed:
chore: add output

fedealconada committed:
chore: fix linter
